skopt/space/space.py

```diff
@@ -104,6 +104,23 @@ def check_dimension(dimension, transform=None):
 
 class Dimension(object):
     """Base class for search space dimensions."""
+    def __init__(self, name=None):
+        """
+        Common initialization for all types of search-spaces dimensions.
+
+        Call this at the start of __init__ for all sub-classes.
+        
+        Parameters
+        ----------
+        * `name` [str or None]:
+            Name associated with the dimension.
+        """
+
+        # Name of this dimension.
+        self.name = name
+
+        # Initialize index before it has been set by search-space.
+        self.index = None
```

Comment:
I still don't like it that we have to reach into a dimension object from the outside to make it "whole". My proposal would be to allow unnamed dimensions `self.name = None` and use `enumerate(dimensions) in the plotting to make up a axis label on the fly. What do others think?

Reply:
I understand your concern. Let me try and explain the reasoning behind this.

The way I chose to resolve the problem of `Categorical` dimensions in the plots, was simply to allow the user to select which dimensions to show by giving a list of dimension-names. This may be useful in other ways as well, if you only want to plot a few of the dimensions for some reason.

Because skopt uses unnamed lists of parameters internally, we need an index for each of the dimensions we are interested in, in order to get the correct parameters from the optimization-log. We cannot use `enumerate(some_dimensions)` because `some_dimensions` are only a subset of `space.dimensions`. So we need to know the indices of `some_dimensions` as they occur in `space.dimensions`.

If the `dimension`-object doesn't know its own index, then we have to look it up from the `space.dimensions` list from the given names of the dimensions. The natural place to have a function for looking up these indices would be in the `space`-object.

To avoid having to go over the `space.dimensions`-list several times; one for looking up the dimensions and another for looking up their indices, I had originally merged all of this into a single function `space.get_dimensions()` in one of the earlier versions of this PR, that would return the relevant `dimension`-objects, their names, and their indices, as I recall. This was the most efficient way of doing it without tampering with your code and without iterating over `space.dimensions` multiple times - but I admitted that it was ugly and as I recall @betatim didn't like it, so I changed it. This was also where @betatim's suggestion for using `space[['dim_name1', 'dim_name2', ...]]` came from, which just returns the `dimension`-objects, but not their indices.

I think the easiest and cleanest way of getting a dimension's index in the search-space, is simply to store it in the `dimension`-object. The only problem with this, is if you want to modify the search-space by adding or remove dimensions, or changing their order. But I don't see a reason for doing that and I don't think it's even possible without mangling the code.

Regarding unnamed dimensions, I think it is a bad idea to give them default names in the plotting functions. Imagine that you have some other functionality some day that also needs default names, then suddenly you are setting default names for the dimensions in different places. I believe it is better to set default names in the `dimension`-object where they belong, and for this they also need to know their indices in the search-space.

If you want to use different names for the plotting-axes e.g. "Learning Rate" instead of "learning_rate" then an idea would be to either set a `display_name` when constructing the `dimension` or giving a list of `display_names` when calling the plotting functions. There are pro's and con's of both approaches.

I hope this was a convincing argument. Otherwise please come up with an alternative design proposal for how we should get these indices.